File: 03-march/25-equal-sum-grid-partition-i/solution.js
var canPartitionGrid = function(grid) {
    const m = grid.length, n = grid[0].length;
    let total = 0;
    for (let i = 0; i < m; i++)
        for (let j = 0; j < n; j++)
            total += grid[i][j];

    if (total % 2 !== 0)
        return false;

    const half = total >> 1;

    let rowSum = 0;
    for (let i = 0; i < m - 1; i++) {
        for (let j = 0; j < n; j++)
            rowSum += grid[i][j];
        if (rowSum === half)
            return true;
    }

    let colSum = 0;
    for (let j = 0; j < n - 1; j++) {
        for (let i = 0; i < m; i++)
            colSum += grid[i][j];
        if (colSum === half)
            return true;
    }

    return false;
};

File: 03-march/25-equal-sum-grid-partition-i/solution.py
class Solution:
    def canPartitionGrid(self, grid: list[list[int]]) -> bool:
        m, n = len(grid), len(grid[0])
        total = sum(sum(row) for row in grid)

        if total % 2 != 0:
            return False

        half = total // 2

        row_sum = 0
        for i in range(m - 1):
            row_sum += sum(grid[i])
            if row_sum == half:
                return True

        col_sum = 0
        for j in range(n - 1):
            for i in range(m):
                col_sum += grid[i][j]
            if col_sum == half:
                return True

        return False
